addons/event_crm/models/event_registration.py

```diff
@@ -0,0 +1,274 @@
+            (registration.id,
+             dict((field, _get_field_value(registration[field], field)) for field in tracked_fields)
+            ) for registration in self
+        )
+
+    def _get_lead_grouping(self, rules, rule_to_new_regs):
+        """ Perform grouping of registrations in order to enable order-based
+        lead creation and update existing groups with new registrations.
+
+        Heuristic in event is the following. Registrations created in multi-mode
+        are grouped by order. Customer use case: website_event use creates several
+        registrations in a create-multi.
+
+        Update is not supported as there is no way to determine if a registration
+        is part of an existing batch.
+
+        :param rules: lead creation rules to run on registrations given by self;
+        :param rule_to_new_regs: dict: for each rule, subset of self matching
+          rule conditions. Used to speedup batch computation;
+
+        :return dict: for each rule, rule (key of dict) gives a list of groups.
+          Each group is a tuple (
+            existing_lead: existing lead to update;
+            group_record: record used to group;
+            registrations: sub record set of self, containing registrations
+                           belonging to the same group;
+          )
+        """
+        event_to_reg_ids = defaultdict(lambda: self.env['event.registration'])
+        for registration in self:
+            event_to_reg_ids[registration.event_id] += registration
+
+        return dict(
+            (rule, [(False, event, registrations & rule_to_new_regs[rule])
+                    for event, registrations in event_to_reg_ids.items()])
+            for rule in rules
+        )
+
+    # ------------------------------------------------------------
+    # TOOLS
+    # ------------------------------------------------------------
+
+    @api.model
+    def _get_lead_contact_fields(self):
+        """ Get registration fields linked to lead contact. Those are used notably
+        to see if an update of lead is necessary or to fill contact values
+        in ``_get_lead_contact_values())`` """
+        return ['name', 'email', 'phone', 'mobile', 'partner_id']
+
+    @api.model
+    def _get_lead_description_fields(self):
+        """ Get registration fields linked to lead description. Those are used
+        notablyto see if an update of lead is necessary or to fill description
+        in ``_get_lead_description())`` """
+        return ['name', 'email', 'phone']
+
+    def _find_first_notnull(self, field_name):
+        """ Small tool to extract the first not nullvalue of a field: its value
+        or the ids if this is a relational field. """
+        value = next((reg[field_name] for reg in self if reg[field_name]), False)
+        if value and self._fields[field_name].type in ('many2many', 'one2many'):
+            return value.ids
+        if value and self._fields[field_name].type == 'many2one':
```

Comment:
Well because it is given to a dict for create / write. It is some kind of convert_to_write, but not sure I should use it so done manually here. I moved it in another small tool, used both time in the code. You can re-check, _convert_value